das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+//
+// Copyright 2022, Offchain Labs, Inc. All rights reserved.
+//
+
+package das
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/bits"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/arbstate"
+	"github.com/offchainlabs/nitro/blsSignatures"
+)
+
+type AggregatorConfig struct {
+	assumedHonest   int
+	retentionPeriod time.Duration
+}
+
+type Aggregator struct {
+	config   AggregatorConfig
+	services []serviceDetails
+
+	/// calculated fields
+	requiredServicesForStore       int
+	maxAllowedServiceStoreFailures int
+}
+
+type serviceDetails struct {
+	service    DataAvailabilityService
+	pubKey     blsSignatures.PublicKey
+	signerMask uint64
+}
+
+func NewAggregator(config AggregatorConfig, services []serviceDetails) *Aggregator {
+	return &Aggregator{
+		config:                         config,
+		services:                       services,
+		requiredServicesForStore:       len(services) + 1 - config.assumedHonest,
+		maxAllowedServiceStoreFailures: config.assumedHonest - 1,
+	}
+}
+
+// Retrieve calls  on each backend DAS in parallel and returns immediately on the
+// first successful response where the data matches the requested hash. Otherwise
+// if all requests fail or if its context is canceled (eg via DeadlineWrapper) then
+// it returns an error.
+func (a *Aggregator) Retrieve(ctx context.Context, cert []byte) ([]byte, error) {
+	requestedCert, _, err := arbstate.DeserializeDASCertFrom(cert)
+	if err != nil {
+		return nil, err
+	}
+
+	// Cert is the aggregate cert, validate it against DAS public keys
+	var servicesThatSignedCert []serviceDetails
+	var pubKeys []blsSignatures.PublicKey
+	for _, d := range a.services {
+		if requestedCert.SignersMask&d.signerMask != 0 {
+			servicesThatSignedCert = append(servicesThatSignedCert, d)
+			pubKeys = append(pubKeys, d.pubKey)
+		}
+	}
+	signedBlob := serializeSignableFields(*requestedCert)
+	sigMatch, err := blsSignatures.VerifySignature(requestedCert.Sig, signedBlob, blsSignatures.AggregatePublicKeys(pubKeys))
+	if err != nil {
+		return nil, err
+	}
+	if !sigMatch {
+		return nil, errors.New("Signature of data in cert passed in doesn't match")
+	}
+
+	blobChan := make(chan []byte)
+	errorChan := make(chan error)
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, d := range servicesThatSignedCert {
+		go func(ctx context.Context, d serviceDetails) {
+			blob, err := d.service.Retrieve(ctx, cert)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+			var blobHash [32]byte
+			copy(blobHash[:], crypto.Keccak256(blob))
+			if blobHash == requestedCert.DataHash {
```

Comment:
Can we use `bytes.Equal` here, to avoid the copy?

Reply:
Fixed, thanks for pointing that out.